src/web/rustdoc.rs

```diff
@@ -549,29 +613,32 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
         "".to_string()
     };
 
-    let query_string = if let Some(query) = req.url.query() {
+    let query_string = if let Some(query) = uri.query() {
         format!("?{}", query)
     } else {
         "".to_string()
     };
 
     let permalink_path = format!(
         "/{}/{}/{}{}",
-        name, latest_version, inner_path, query_string
+        params.name, latest_version, inner_path, query_string
     );
 
-    let latest_path = format!("/crate/{}/latest{}{}", name, target_redirect, query_string);
+    let latest_path = format!(
+        "/crate/{}/latest{}{}",
+        params.name, target_redirect, query_string
+    );
 
     // Set the canonical URL for search engines to the `/latest/` page on docs.rs.
     // Note: The URL this points to may not exist. For instance, if we're rendering
     // `struct Foo` in version 0.1.0 of a crate, and version 0.2.0 of that crate removes
     // `struct Foo`, this will point at a 404. That's fine: search engines will crawl
     // the target and will not canonicalize to a URL that doesn't exist.
     // Don't include index.html in the canonical URL.
-    let canonical_url = format!(
-        "https://docs.rs/{}/latest/{}",
-        name,
-        inner_path.replace("index.html", "")
+    let canonical_path = format!(
+        "/{}/latest/{}",
+        utf8_percent_encode(&params.name, CONTROLS),
+        utf8_percent_encode(&inner_path.replace("index.html", ""), CONTROLS),
```

Comment:
why do we need to manually encode this?

Reply:
So the canonical URI will be a valid URI with only valid URI characters. 

I had errors coming from the new header encoding because it started validating the URI, coming from filenames with UTF8 characters.

Reply:
See also https://github.com/rust-lang/docs.rs/pull/1968